### PR TITLE
Enforce, fix, and modernize linting

### PR DESCRIPTION
**Pull request by bakkot**
`npm run lint` existed, but was unused (and broken). This PR adds it to `npm run test` so it will get enforced by CI, fixes it, adds linting to typescript files, and bumps eslint.

I haven't changed any of the linting options (at least, not intentionally), though there's some options I should certainly like to add. But we can bikeshed those in another issue or PR.

### CODE REVIEW

File: bin/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  extends: '../.eslintrc.js',
```
Comment (ljharb) [inline bin/.eslintrc.js]: this should be done with "overrides" in the root, rather than having more than one eslint config file.

File: .eslintrc [hunk 2]
```diff
@@ -1,45 +0,0 @@
-{
```
Comment (ljharb) [inline .eslintrc]: i would very much prefer to retain the `.eslintrc` format rather than using JS for config

Comment (bakkot) [reply .eslintrc]: Moved to `.eslintrc.json`.

File: bin/.eslintrc.js (hunk 1 above)
Comment (bakkot) [reply bin/.eslintrc.js]: Done.

**Pull request merged**